packages/exhibitor/extra/dcos_zk_backup.py

```diff
@@ -0,0 +1,350 @@
+import shutil
+import subprocess
+import sys
+import tarfile
+from argparse import ArgumentParser, ArgumentTypeError
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import Optional
+
+
+EXHIBITOR_DIR = Path('/var/lib/dcos/exhibitor')
+
+
+def run_command(cmd: str, verbose: bool) -> None:
+    """
+    Run a command in a subprocess.
+
+    Args:
+        verbose: Show the output.
+
+    Raises:
+        subprocess.CalledProcessError: The given cmd exits with a non-0 exit code.
+    """
+    stdout = None if verbose else subprocess.PIPE
+    stderr = None if verbose else subprocess.STDOUT
+    subprocess.run(
+        args=shlex.split(cmd),
+        stdout=stdout,
+        stderr=stderr,
+        check=True,
+    )
+
+
+def _is_zookeeper_running(verbose: bool) -> bool:
+    """
+    Exhibitor and therefore ZooKeeper is in active state (running).
+    """
+    try:
+        run_command('systemctl status dcos-exhibitor', verbose)
+    except subprocess.CalledProcessError:
+        # Non-zero exit code indicates Exhibitor + ZooKeeper are dead.
+        return False
+    return True
+
+
+def _copy_dir_and_preserve_ownership(src: Path, dst: Path, verbose: bool) -> None:
+    """
+    Copy a directory from ``src`` to ``dst`` and preserve ownership.
+
+    We need to preserve ownership for a successful restore.
+    We use a subprocess (with `cp -p`) for this rather than Python's
+    `shutil.copytree` as `copytree` does not preserve ownership.
+    """
+    run_command('cp -prv {src} {dst}'.format(src=src, dst=dst), verbose)
+
+
+def backup_zookeeper(
+    backup: Path,
+    tmp_dir: Path,
+    verbose: bool,
+) -> None:
+    """
+    DC/OS ZooKeeper instance backup procedure.
+
+    ZooKeeper changes files while it is running.  In order to have a consistent
+    backup stop ZooKeeper, run the backup procedure, then start ZooKeeper.
+
+    See https://jira.mesosphere.com/browse/DCOS_OSS-5185
+    for changing this procedure to allow a backup without downtime.
+
+    If backing up ZooKeeper via a custom temporary directory, a failing backup
+    procedure does not clean up files created in the temporary directory in the
+    process.
+    """
+    zookeeper_dir = EXHIBITOR_DIR / 'zookeeper'
+    tmp_zookeeper_dir = tmp_dir / 'zookeeper'
+
+    print('Backing up ZooKeeper into {backup} via {tmp_zookeeper_dir}'.format(
+        backup=backup,
+        tmp_zookeeper_dir=tmp_zookeeper_dir,
+    ))
+
+    print('Validate that ZooKeeper is not running')
+    assert not _is_zookeeper_running(verbose)
+
+    print('Copying ZooKeeper files to {tmp_zookeeper_dir}'.format(
+        tmp_zookeeper_dir=tmp_zookeeper_dir,
+    ))
+
+    _copy_dir_and_preserve_ownership(
+        src=zookeeper_dir,
+        dst=tmp_zookeeper_dir,
+        verbose=verbose,
+    )
+
+    print('Creating ZooKeeper backup tar archive at {backup}'.format(backup=backup))
+
+    def _tar_filter(tar_info: tarfile.TarInfo) -> Optional[tarfile.TarInfo]:
```

Comment:
Could you add a comment explaining why the files which match these names are not welcome in the backup?